src/Microsoft.OData.Core/ODataBatchReader.cs

```diff
@@ -293,45 +409,76 @@ private bool ReadImplementation()
-                    this.operationState = OperationState.None;
+                    this.ReaderOperationState = OperationState.None;
 
                     // Also add a pending ContentId header to the URL resolver now. We ensured above
                     // that a message has been created for this operation and thus the headers (incl.
                     // a potential content ID header) have been read.
                     if (this.contentIdToAddOnNextRead != null)
                     {
+                        if (this.payloadUriConverter.ContainsContentId(this.contentIdToAddOnNextRead))
+                        {
+                            throw new ODataException(
+                                Strings.ODataBatchReader_DuplicateContentIDsNotAllowed(this.contentIdToAddOnNextRead));
+                        }
+
                         this.payloadUriConverter.AddContentId(this.contentIdToAddOnNextRead);
                         this.contentIdToAddOnNextRead = null;
                     }
 
                     // When we are done with an operation, we have to skip ahead to the next part
                     // when Read is called again. Note that if the operation stream was never requested
                     // and the content of the operation has not been read, we'll skip it here.
-                    Debug.Assert(this.operationState == OperationState.None, "Operation state must be 'None' at the end of the operation.");
-                    this.batchReaderState = this.SkipToNextPartAndReadHeaders();
+                    Debug.Assert(this.ReaderOperationState == OperationState.None, "Operation state must be 'None' at the end of the operation.");
+
+                    if (this.isInChangeset)
+                    {
+                        this.IncreaseChangesetSize();
+                    }
+                    else
+                    {
+                        this.IncreaseBatchSize();
+                    }
+
+                    this.State = this.ReadAtOperationImplementation();
+
                     break;
 
                 case ODataBatchReaderState.ChangesetStart:
                     // When at the start of a changeset, skip ahead to the first operation in the
                     // changeset (or the end boundary of the changeset).
-                    Debug.Assert(this.batchStream.ChangeSetBoundary != null, "Changeset boundary must have been set by now.");
-                    this.batchReaderState = this.SkipToNextPartAndReadHeaders();
+                    if (this.isInChangeset)
+                    {
+                        ThrowODataException(Strings.ODataBatchReaderStream_NestedChangesetsAreNotSupported);
+                    }
+
+                    // Increment the batch size at the start of the changeset since we haven't counted it yet
+                    // when the state was transitioned from Operation upon detection of this sub-batch.
+                    this.IncreaseBatchSize();

```

Comment:
If the changeset is the first thing in the batch, won't this result in an extra count in the batchsize? And, don't we go back to changesetstart between each operation in the changeset?

Reply:
ODataBatchReaderState.ChangesetStart is reached when the changeset preamble is parsed/detected, therefore it won't result in double counting.
Within a changeset, transition from one operation to another won't result in transient changesetstart state. Please note that nested changeset is not allowed, otherwise it would be the case.

---
In reply to: [145849860](https://github.com/OData/odata.net/pull/960#discussion_r145849860) [](ancestors = 145849860)

Reply:
Okay; let's be sure we have a test to test that the counts are correct for the changest being the first thing in the batch as well as changeset coming after an operation.

---
In reply to: [146066241](https://github.com/OData/odata.net/pull/960#discussion_r146066241) [](ancestors = 146066241,145849860)